teps/0115-tekton-catalog-git-based-versioning.md

```diff
@@ -0,0 +1,680 @@
+
+## Definitions
+
+Terms used in this TEP are defined as follows:
+
+* **Repository**: A versioned source of a project. There's no special definition of "repository" in the context of this
+TEP. A repository may be a Catalog.
+
+* **Catalog**: A repository that complies with the organization contract defined by Tekton. The directory-based contract
+is defined in [TEP-0003: Tekton Catalog Organization][tep-0003-org], and [git-based contract](#organization-contract) is
+defined in this TEP. A Catalog may contain one or more resources.
+
+* **Resource**: Item shared in a Catalog e.g. `Task` or `Pipeline`.
+
+* **Hub**: User interface for one or more Catalogs.
+
+## Proposal
+
+We propose supporting git-based versioning of Tekton Catalogs.
+
+### Git-Based Versioning
+
+#### Organization Contract
+
+We propose removing the directories - `0.1`...`0.n` - that were used for versioning resources in the Catalogs. The new
+organization contract with git-based versioning would be:
+
+```bash
+# Optional: owner(s) and reviewer(s) of the resource (all versions)
+./{resource-type}/{resource-name}/OWNERS
+# The README of the resource (what is the resource about, usage, ...)
+./{resource-type}/{resource-name}/README.md
+
+# The resource itself
+./{resource-type}/{resource-name}/{resource-name}.yaml
+
+# Optional: Tests verifying functionality.
+./{resource-type}/{resource-name}/tests/...
+
+# Optional: Additional samples.
+./{resource-type}/{resource-name}/{version}/samples/... |
+```
+
+The example directory structure shown [above](#motivation) will be updated as such:
+
+```bash
+./task/
+  /argocd
+    /README.md
+    /argocd.yaml
+    /samples
+    /tests
+    /OWNERS
+  /golang-build
+    /README.md
+    /golang-build.yaml
+    /samples
+    /tests
+    /OWNERS
+  /foo-task
+  /bar-task
+  /...
+./pipeline/
+  /go-release
+    /README.md
+    /go-release.yaml
+    /samples
+    /tests
+    /OWNERS
+  /foo-pipeline
+  /bar-pipeline
+  /...  
+```
+
+Note that there are no changes to the contents of the resources themselves, only the organization contract of Catalogs.
+
+#### Versioning in a Catalog
+
+Versioning will move from the resources themselves to the Catalog as a whole. This is because releases and version tags
+apply to the whole repository. Therefore, Catalogs will contain one or multiple resource with shared versioning.
+
+The versions in the Tekton Hub should continue to use the simple version of [semver][semver]: `{major}.{minor}`. We
+propose that the release tags in Catalogs use the semantic versioning model - `v<major-version>.<minor-version>`. 
+Minor versions are made for backwards-compatible changes, while major versions are made for backwards-incompatible
+changes. As such, the tags in the Catalogs will map to versions in the Hub e.g. tag "v0.7" in a Catalog maps to version
+"0.7" in the Hub. The Hub will recognize this versioning model only, therefore Catalogs must use this versioning to
+surface versions in the Hub.
```

Comment:
This model is better than the existing one, and I think this model works well with one resource (or a coherent set of resources) in the catalog. It's not clear how it should be applied to a catalog like the one we have today, where you have multiple unrelated resources and multiple unrelated owners. 

This difficulty in versioning a catalog with multiple unrelated resources in it is the main reason the old model was chosen, and if we want to change that we need to provide a strong alternative. 

If the only alternative is to break down the catalog in many many repos we need to state that clearly, and provide more guidance about how to do it, because there is a non-trivial overhead in setting up all the different repos, their CI, CD, owners etc.

If we still consider hosting many unrelated resources in a single repo, we should state so and provide an example about how versioning would work in that case.

One option for those catalogs could be periodic releases and the version number that does not follow semantic versioning, so users must rely on release notes to discover whether breaking changes exists and where.

Reply:
added more details in the TEP about releases as discussed earlier today 😁